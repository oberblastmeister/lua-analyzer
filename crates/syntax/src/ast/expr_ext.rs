use crate::{
    ast::{self, support, AstNode, AstToken},
    SyntaxToken, T,
};
use parser::LuaOp;

impl ast::InfixExpr {
    pub fn op_details(&self) -> Option<(SyntaxToken, LuaOp)> {
        self.syntax()
            .children_with_tokens()
            .filter_map(|it| it.into_token())
            .find_map(|c| {
                let kind = c.kind();
                if kind == T![not] {
                    None
                } else {
                    <Option<LuaOp>>::from(kind).map(|op| (c, op))
                }
            })
    }

    pub fn op_kind(&self) -> Option<LuaOp> {
        self.op_details().map(|t| t.1)
    }

    pub fn op_token(&self) -> Option<SyntaxToken> {
        self.op_details().map(|t| t.0)
    }

    pub fn lhs(&self) -> Option<ast::Expr> {
        self.sub_exprs().0
    }

    pub fn rhs(&self) -> Option<ast::Expr> {
        self.sub_exprs().1
    }

    pub fn sub_exprs(&self) -> (Option<ast::Expr>, Option<ast::Expr>) {
        let mut children = support::children(self.syntax());
        let first = children.next();
        let second = children.next();
        (first, second)
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Hash)]
pub enum LiteralKind {
    Str(ast::Str),
    Number(ast::Number),
    Bool(bool),
    Nil,
    Vararg,
}

impl ast::Literal {
    pub fn token(&self) -> SyntaxToken {
        self.syntax()
            .children_with_tokens()
            .find(|e| !e.kind().is_trivia())
            .and_then(|e| e.into_token())
            .unwrap()
    }

    pub fn kind(&self) -> LiteralKind {
        let token = self.token();

        if let Some(t) = ast::Number::cast(token.clone()) {
            return LiteralKind::Number(t);
        }

        if let Some(t) = ast::Str::cast(token.clone()) {
            return LiteralKind::Str(t);
        }

        match token.kind() {
            T![true] => LiteralKind::Bool(true),
            T![false] => LiteralKind::Bool(false),
            T![nil] => LiteralKind::Nil,
            T![...] => LiteralKind::Vararg,
            _ => unreachable!(),
        }
    }
}
