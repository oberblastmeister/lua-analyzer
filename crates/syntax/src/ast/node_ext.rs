use crate::{ast, SyntaxNode, TokenText};

use super::AstNode;

impl ast::Name {
    pub fn text(&self) -> TokenText {
        text_of_first_token(self.syntax())
    }
}

impl ast::NameRef {
    pub fn text(&self) -> TokenText {
        text_of_first_token(self.syntax())
    }
}

fn text_of_first_token(node: &SyntaxNode) -> TokenText {
    let first_token =
        node.green().children().next().and_then(|it| it.into_token()).unwrap().to_owned();

    TokenText(first_token)
}
