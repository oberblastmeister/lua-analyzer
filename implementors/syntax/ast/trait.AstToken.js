(function() {var implementors = {};
implementors["syntax"] = [];
if (window.register_implementors) {window.register_implementors(implementors);} else {window.pending_implementors = implementors;}})()