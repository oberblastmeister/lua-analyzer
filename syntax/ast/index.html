<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="API documentation for the Rust `ast` mod in crate `syntax`."><meta name="keywords" content="rust, rustlang, rust-lang, ast"><title>syntax::ast - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings"></script><script src="../../storage.js"></script><script src="../../crates.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="icon" type="image/svg+xml" href="../../favicon.svg">
<link rel="alternate icon" type="image/png" href="../../favicon-16x16.png">
<link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 8]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../../syntax/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><p class="location">Module ast</p><div class="sidebar-elems"><div class="block items"><ul><li><a href="#structs">Structs</a></li><li><a href="#enums">Enums</a></li><li><a href="#traits">Traits</a></li></ul></div><p class="location"><a href="../index.html">syntax</a></p><div id="sidebar-vars" data-name="ast" data-ty="mod" data-relpath="../"></div><script defer src="../sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu"><img src="../../brush.svg" width="18" height="18" alt="Pick another theme!"></button><div id="theme-choices" role="menu"></div></div><script src="../../theme.js"></script><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" disabled autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" class="help-button">?</button>
                <a id="settings-menu" href="../../settings.html"><img src="../../wheel.svg" width="18" height="18" alt="Change settings"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Module <a href="../index.html">syntax</a>::<wbr><a class="mod" href="">ast</a></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/syntax/ast.rs.html#1-92" title="goto source code">[src]</a></span></h1><h2 id="structs" class="section-header"><a href="#structs">Structs</a></h2>
<table><tr class="module-item"><td><a class="struct" href="struct.ArgList.html" title="syntax::ast::ArgList struct">ArgList</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.AssignStmt.html" title="syntax::ast::AssignStmt struct">AssignStmt</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.AstChildren.html" title="syntax::ast::AstChildren struct">AstChildren</a></td><td class="docblock-short"><p>An iterator over <code>SyntaxNode</code> children of a particular AST type.</p>
</td></tr><tr class="module-item"><td><a class="struct" href="struct.Block.html" title="syntax::ast::Block struct">Block</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.BreakStmt.html" title="syntax::ast::BreakStmt struct">BreakStmt</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.CallArgs.html" title="syntax::ast::CallArgs struct">CallArgs</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.CallExpr.html" title="syntax::ast::CallExpr struct">CallExpr</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.Comment.html" title="syntax::ast::Comment struct">Comment</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.DoStmt.html" title="syntax::ast::DoStmt struct">DoStmt</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.DotExpr.html" title="syntax::ast::DotExpr struct">DotExpr</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.ElseBranch.html" title="syntax::ast::ElseBranch struct">ElseBranch</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.ElseIfBranch.html" title="syntax::ast::ElseIfBranch struct">ElseIfBranch</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.Error.html" title="syntax::ast::Error struct">Error</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.ExprStmt.html" title="syntax::ast::ExprStmt struct">ExprStmt</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.False.html" title="syntax::ast::False struct">False</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.ForStmt.html" title="syntax::ast::ForStmt struct">ForStmt</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.FunctionDefStmt.html" title="syntax::ast::FunctionDefStmt struct">FunctionDefStmt</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.FunctionExpr.html" title="syntax::ast::FunctionExpr struct">FunctionExpr</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.FunctionMethod.html" title="syntax::ast::FunctionMethod struct">FunctionMethod</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.FunctionStatic.html" title="syntax::ast::FunctionStatic struct">FunctionStatic</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.GenericFor.html" title="syntax::ast::GenericFor struct">GenericFor</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.GotoStmt.html" title="syntax::ast::GotoStmt struct">GotoStmt</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.Ident.html" title="syntax::ast::Ident struct">Ident</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.IdentKey.html" title="syntax::ast::IdentKey struct">IdentKey</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.IfStmt.html" title="syntax::ast::IfStmt struct">IfStmt</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.Index.html" title="syntax::ast::Index struct">Index</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.IndexExpr.html" title="syntax::ast::IndexExpr struct">IndexExpr</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.IndexPath.html" title="syntax::ast::IndexPath struct">IndexPath</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.InfixExpr.html" title="syntax::ast::InfixExpr struct">InfixExpr</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.KeyValue.html" title="syntax::ast::KeyValue struct">KeyValue</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.LabelDelim.html" title="syntax::ast::LabelDelim struct">LabelDelim</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.LabelStmt.html" title="syntax::ast::LabelStmt struct">LabelStmt</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.Literal.html" title="syntax::ast::Literal struct">Literal</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.LocalAssignStmt.html" title="syntax::ast::LocalAssignStmt struct">LocalAssignStmt</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.LocalFunctionDefStmt.html" title="syntax::ast::LocalFunctionDefStmt struct">LocalFunctionDefStmt</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.MethodCallExpr.html" title="syntax::ast::MethodCallExpr struct">MethodCallExpr</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.MultiName.html" title="syntax::ast::MultiName struct">MultiName</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.MultivalExpr.html" title="syntax::ast::MultivalExpr struct">MultivalExpr</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.Name.html" title="syntax::ast::Name struct">Name</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.NameRef.html" title="syntax::ast::NameRef struct">NameRef</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.Number.html" title="syntax::ast::Number struct">Number</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.NumericFor.html" title="syntax::ast::NumericFor struct">NumericFor</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.ParamList.html" title="syntax::ast::ParamList struct">ParamList</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.ParenExpr.html" title="syntax::ast::ParenExpr struct">ParenExpr</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.PositionalValue.html" title="syntax::ast::PositionalValue struct">PositionalValue</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.PrefixExpr.html" title="syntax::ast::PrefixExpr struct">PrefixExpr</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.RepeatUntilStmt.html" title="syntax::ast::RepeatUntilStmt struct">RepeatUntilStmt</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.ReturnStmt.html" title="syntax::ast::ReturnStmt struct">ReturnStmt</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.Shebang.html" title="syntax::ast::Shebang struct">Shebang</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.SourceFile.html" title="syntax::ast::SourceFile struct">SourceFile</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.Str.html" title="syntax::ast::Str struct">Str</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.TableExpr.html" title="syntax::ast::TableExpr struct">TableExpr</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.TableSep.html" title="syntax::ast::TableSep struct">TableSep</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.True.html" title="syntax::ast::True struct">True</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.WhileStmt.html" title="syntax::ast::WhileStmt struct">WhileStmt</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="struct" href="struct.Whitespace.html" title="syntax::ast::Whitespace struct">Whitespace</a></td><td class="docblock-short"></td></tr></table><h2 id="enums" class="section-header"><a href="#enums">Enums</a></h2>
<table><tr class="module-item"><td><a class="enum" href="enum.Expr.html" title="syntax::ast::Expr enum">Expr</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="enum" href="enum.ForContent.html" title="syntax::ast::ForContent enum">ForContent</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="enum" href="enum.FunctionDefContent.html" title="syntax::ast::FunctionDefContent enum">FunctionDefContent</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="enum" href="enum.LiteralKind.html" title="syntax::ast::LiteralKind enum">LiteralKind</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="enum" href="enum.Stmt.html" title="syntax::ast::Stmt enum">Stmt</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="enum" href="enum.StmtItem.html" title="syntax::ast::StmtItem enum">StmtItem</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="enum" href="enum.TableContent.html" title="syntax::ast::TableContent enum">TableContent</a></td><td class="docblock-short"></td></tr><tr class="module-item"><td><a class="enum" href="enum.TableKey.html" title="syntax::ast::TableKey enum">TableKey</a></td><td class="docblock-short"></td></tr></table><h2 id="traits" class="section-header"><a href="#traits">Traits</a></h2>
<table><tr class="module-item"><td><a class="trait" href="trait.AstNode.html" title="syntax::ast::AstNode trait">AstNode</a></td><td class="docblock-short"><p>The main trait to go from untyped <code>SyntaxNode</code>  to a typed ast. The
conversion itself has zero runtime cost: ast and syntax nodes have exactly
the same representation: a pointer to the tree root and a pointer to the
node itself.</p>
</td></tr><tr class="module-item"><td><a class="trait" href="trait.AstToken.html" title="syntax::ast::AstToken trait">AstToken</a></td><td class="docblock-short"><p>Like <code>AstNode</code>, but wraps tokens rather than interior nodes.</p>
</td></tr></table></section><section id="search" class="content hidden"></section><section class="footer"></section><div id="rustdoc-vars" data-root-path="../../" data-current-crate="syntax" data-search-js="../../search-index.js"></div>
    <script src="../../main.js"></script></body></html>