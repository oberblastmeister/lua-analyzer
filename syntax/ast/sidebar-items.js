initSidebarItems({"enum":[["Expr",""],["ForContent",""],["FunctionDefContent",""],["LiteralKind",""],["Stmt",""],["StmtItem",""],["TableContent",""],["TableKey",""]],"struct":[["ArgList",""],["AssignStmt",""],["AstChildren","An iterator over `SyntaxNode` children of a particular AST type."],["Block",""],["BreakStmt",""],["CallArgs",""],["CallExpr",""],["Comment",""],["DoStmt",""],["DotExpr",""],["ElseBranch",""],["ElseIfBranch",""],["Error",""],["ExprStmt",""],["False",""],["ForStmt",""],["FunctionDefStmt",""],["FunctionExpr",""],["FunctionMethod",""],["FunctionStatic",""],["GenericFor",""],["GotoStmt",""],["Ident",""],["IdentKey",""],["IfStmt",""],["Index",""],["IndexExpr",""],["IndexPath",""],["InfixExpr",""],["KeyValue",""],["LabelDelim",""],["LabelStmt",""],["Literal",""],["LocalAssignStmt",""],["LocalFunctionDefStmt",""],["MethodCallExpr",""],["MultiName",""],["MultivalExpr",""],["Name",""],["NameRef",""],["Number",""],["NumericFor",""],["ParamList",""],["ParenExpr",""],["PositionalValue",""],["PrefixExpr",""],["RepeatUntilStmt",""],["ReturnStmt",""],["Shebang",""],["SourceFile",""],["Str",""],["TableExpr",""],["TableSep",""],["True",""],["WhileStmt",""],["Whitespace",""]],"trait":[["AstNode","The main trait to go from untyped `SyntaxNode`  to a typed ast. The conversion itself has zero runtime cost: ast and syntax nodes have exactly the same representation: a pointer to the tree root and a pointer to the node itself."],["AstToken","Like `AstNode`, but wraps tokens rather than interior nodes."]]});