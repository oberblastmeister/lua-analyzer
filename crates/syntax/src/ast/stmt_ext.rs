use crate::ast::{self, support, AstNode};

impl ast::Stmt {
    pub fn is_block_like(&self) -> bool {
        matches!(
            self,
            ast::Stmt::IfStmt(_)
                | ast::Stmt::ForStmt(_)
                | ast::Stmt::WhileStmt(_)
                | ast::Stmt::FunctionDefStmt(_)
                | ast::Stmt::RepeatUntilStmt(_)
        )
    }
}

impl ast::AssignStmt {
    pub fn lhs(&self) -> Option<ast::MultivalExpr> {
        self.sub_multival_exprs().0
    }

    pub fn rhs(&self) -> Option<ast::MultivalExpr> {
        self.sub_multival_exprs().1
    }

    pub fn sub_multival_exprs(&self) -> (Option<ast::MultivalExpr>, Option<ast::MultivalExpr>) {
        let mut children = support::children(self.syntax());
        let first = children.next();
        let second = children.next();
        (first, second)
    }
}
