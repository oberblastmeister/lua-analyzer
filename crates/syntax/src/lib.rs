pub mod ast;
mod parsing;
mod syntax_node;

use std::{marker::PhantomData, sync::Arc};

use ast::AstNode;
pub use parser::{SyntaxKind, T};
use rowan::GreenNode;

pub use syntax_node::{
    SyntaxElement, SyntaxElementChildren, SyntaxError, SyntaxNode, SyntaxNodeChildren, SyntaxToken,
};

/// `Parse` is the result of the parsing: a syntax tree and a collection of
/// errors.
///
/// Note that we always produce a syntax tree, even for completely invalid
/// files.
#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Parse<T> {
    green: GreenNode,
    errors: Arc<Vec<SyntaxError>>,
    _ty: PhantomData<fn() -> T>,
}

impl<T> Parse<T> {
    fn new(green: GreenNode, errors: Vec<SyntaxError>) -> Parse<T> {
        Parse {
            green,
            errors: Arc::new(errors),
            _ty: PhantomData,
        }
    }

    pub fn syntax_node(&self) -> SyntaxNode {
        SyntaxNode::new_root(self.green.clone())
    }
}

impl<T: AstNode> Parse<T> {
    pub fn to_syntax(self) -> Parse<SyntaxNode> {
        Parse {
            green: self.green,
            errors: self.errors,
            _ty: PhantomData,
        }
    }

    pub fn tree(&self) -> T {
        T::cast(self.syntax_node()).unwrap()
    }

    pub fn errors(&self) -> &[SyntaxError] {
        &*self.errors
    }

    pub fn ok(self) -> Result<T, Arc<Vec<SyntaxError>>> {
        if self.errors.is_empty() {
            Ok(self.tree())
        } else {
            Err(self.errors)
        }
    }
}

impl Parse<SyntaxNode> {
    pub fn cast<N: AstNode>(self) -> Option<Parse<N>> {
        if N::cast(self.syntax_node()).is_some() {
            Some(Parse {
                green: self.green,
                errors: self.errors,
                _ty: PhantomData,
            })
        } else {
            None
        }
    }
}

use ast::Program;

impl Program {
    pub fn parse(text: &str) -> Parse<Program> {
        let (green, errors) = parsing::parse_text(text);
        let root = SyntaxNode::new_root(green.clone());

        assert_eq!(root.kind(), SyntaxKind::Program);
        Parse {
            green,
            errors: Arc::new(errors),
            _ty: PhantomData,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use core::fmt;
    use std::fs;

    fn format_errors(errors: &[SyntaxError]) -> String {
        let mut s = String::new();
        s.push_str(
            "
=============================
Errors:
=============================
",
        );
        s.push_str(&format!("{:#?}", errors));
        s
    }

    fn dump_parse<T: AstNode + fmt::Debug>(parse: Parse<T>) -> String {
        let mut s = String::new();
        s.push_str(&format!("{:#?}", parse.syntax_node()));
        s.push_str(
            "
=============================
Errors:
=============================
",
        );
        s.push_str(&format!("{:#?}", parse.errors()));
        s
    }

    fn dump_parse_no_errors<T: AstNode + fmt::Debug>(parse: Parse<T>) -> String {
        let s = format!("{:#?}", parse.syntax_node());
        if !parse.errors().is_empty() {
            panic!(
                "Should not have any errors {}",
                format_errors(parse.errors())
            )
        }
        s
    }

    macro_rules! get_insta_path {
        ($dir:literal, $name:ident, $closure:expr) => {{
            let single_path_glob =
                concat!("snapshot_inputs/", $dir, "/", stringify!($name), ".lua");
            insta::glob!(single_path_glob, $closure)
        }};
    }

    macro_rules! do_test {
        ("successes", $name:ident) => {
            #[test]
            fn $name() {
                get_insta_path!("successes", $name, |path| {
                    use std::fs;

                    let input = fs::read_to_string(path)
                        .unwrap_or_else(|_| panic!("Failed to read path {} to a string", path.display()));
                    let suffix = path.file_stem().unwrap().to_str().unwrap();

                    insta::with_settings!(
                        {
                            snapshot_suffix => suffix,
                            snapshot_path => "snapshots/successes",
                        },
                        {
                            insta::assert_snapshot!(dump_parse_no_errors(Program::parse(&input)))
                        }
                    )
                })
            }
        };
        ("fails", $name:ident) => {
            #[test]
            fn $name() {
                get_insta_path!("fails", $name, |path| {
                    use std::fs;

                    let input = fs::read_to_string(path)
                        .unwrap_or_else(|_| panic!("Failed to read path {} to a string", path.display()));
                    let suffix = path.file_stem().unwrap().to_str().unwrap();

                    insta::with_settings!(
                        {
                            snapshot_suffix => suffix,
                            snapshot_path => "snapshots/fails",
                        },
                        {
                            insta::assert_snapshot!(dump_parse(Program::parse(&input)))
                        }
                    )
                })
            }
        }
    }

    macro_rules! test_successes {
        ($($name:ident),+ $(,)?) => {
            $(
                do_test!("successes", $name);
             )+
        };
    }

    macro_rules! test_fails {
        ($($name:ident),+ $(,)?) => {
            $(
                do_test!("fails", $name);
             )+
        };
    }

    test_successes![
        assign,
        comments,
        function_call,
        function_call_multi,
        function_def,
        function_literal,
        higher_order_function,
        simple_math,
    ];

    test_fails![can_call, cannot_call_literal, missing_paren];

    #[test]
    fn nothing() {
        insta::assert_debug_snapshot!(Program::parse("").syntax_node(), @"Program@0..0
");
    }
}
