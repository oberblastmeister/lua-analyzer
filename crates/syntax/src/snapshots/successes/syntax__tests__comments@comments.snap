---
source: crates/syntax/src/lib.rs
expression: "dump_parse_no_errors(Program::parse(&input))"

---
Program@0..68
  Comment@0..34 "-- these are just sim ..."
  Comment@34..51 "-- more comments\n"
  Comment@51..68 "-- more comments\n"

